pkg/apis/pipeline/v1alpha1/pipeline_validation.go

```diff
@@ -128,19 +128,35 @@ func (ps *PipelineSpec) Validate(ctx context.Context) *apis.FieldError {
 	// Names cannot be duplicated
 	taskNames := map[string]struct{}{}
 	for i, t := range ps.Tasks {
-		// Task names are appended to the container name, which must exist and
-		// must be a valid k8s name
-		if errSlice := validation.IsQualifiedName(t.Name); len(errSlice) != 0 {
-			return apis.ErrInvalidValue(strings.Join(errSlice, ","), fmt.Sprintf("spec.tasks[%d].name", i))
+		// can't have both taskRef and taskSpec at the same time
+		if (t.TaskRef != nil && t.TaskRef.Name != "") && t.TaskSpec != nil {
+			return apis.ErrMultipleOneOf(fmt.Sprintf("spec.tasks[%d].taskRef", i), fmt.Sprintf("spec.tasks[%d].taskSpec", i))
 		}
-		// TaskRef name must be a valid k8s name
-		if errSlice := validation.IsQualifiedName(t.TaskRef.Name); len(errSlice) != 0 {
-			return apis.ErrInvalidValue(strings.Join(errSlice, ","), fmt.Sprintf("spec.tasks[%d].taskRef.name", i))
+		// Check that one of TaskRef and TaskSpec is present
+		if (t.TaskRef == nil || (t.TaskRef != nil && t.TaskRef.Name == "")) && t.TaskSpec == nil {
+			return apis.ErrMissingOneOf(fmt.Sprintf("spec.tasks[%d].taskRef", i), fmt.Sprintf("spec.tasks[%d].taskSpec", i))
 		}
-		if _, ok := taskNames[t.Name]; ok {
-			return apis.ErrMultipleOneOf(fmt.Sprintf("spec.tasks[%d].name", i))
+		// Validate TaskSpec if it's present
+		if t.TaskSpec != nil {
+			if err := t.TaskSpec.Validate(ctx); err != nil {
+				return err
+			}
```

Comment:
NIT: I guess we could return nil here, since we have a spec so we cannot have a TaskRef and we checked for that...

Reply:
line 142 returns an error from the taskSpec validation, if a task has `taskSpec`, validate the specification and return error if validation fails.

Reply:
@pritidesai it's a nit but I tend to be on @afrittoli side in that, if the `TaskSpec` is valid, we can safely `return nil` and do not care about the rest of the function. As of the code is done today, the rest of the function will be no-op, but I tend to prefer returning early and explicitely

Reply:
haha I am confused now 😢 `return err` on line 142 returns `error` if `taskSpec` validation fails otherwise continues to next task in a `for` loop. you saying `return  err` could be replaced with `return nil` 🤔 or add `return nil` after line 144?

Reply:
@pritidesai sorry about that :stuck_out_tongue: Also, just realized it's in a loop, so the `return nil` wouldn't be correct :stuck_out_tongue:. We could extract that `validate` part on its own, but we can do that later on :wink: so, forget I ever said anything :bowing_man: :joy_cat: :stuck_out_tongue_closed_eyes:

Reply:
Oh, yeah, my bad, sorry about the confusion :(

Reply:
no worries @afrittoli 👍